Refactoring runtime controller and adding new Knative Build

venturasr opened the pull request:
This merge requests implements the knative function controller for knative version `0.6.1`.

Details
- adds function status
- uses knative 0.6.1

ToDo:

- [x] Fix existing Tests
- [x] Uncomment commented Tests
  - [x] `serving_utils_test.go`
- [x] Make status working
- [x] Get controller working
- [x] create, delete, create function => function is not handled properly
  - no function status at all
  - no function knative resources get created
- [x] [Use kustomize and RBAC generation from code](https://github.com/kyma-incubator/runtime/pull/5)
- [x] make webhook work again

nachtmaar committed:
Add mutating webhook for function CRD

nachtmaar committed:
Add more comments to tests

santiago-ventura committed:
Refactor runtime controller and adding new Knative Build

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/kyma-incubator/runtime?pullRequest=7) <br/>All committers have signed the CLA.

santiago-ventura committed:
README updated for Knative version v0.6

nachtmaar committed:
Add sed patching of kustomize patch

santiago-ventura committed:
- Add new columns to crd
- Add a new config yaml for testing
- Update README-md
- Remove no nedeed crd for serving

santiago-ventura committed:
Add RBC for build in function controller
Add namespaces in config file.
Add namespace env in manager.yaml

santiago-ventura committed:
Add namespace to configmaps

nachtmaar committed:
4223 implement webhook for knative function crd integration test (#8)

Add integration test for webhook server

santiago-ventura committed:
Add FunctionStatus attr in function_types.go Update the status of the function

nachtmaar committed:
Merge remote-tracking branch 'origin/master' into refactored-runtime

nachtmaar committed:
Fix typos

nachtmaar committed:
Add createFunctionHandlerMap test

nachtmaar commented:
I added some todos to the description such that we can better keep track of that needs to be done/what is already done.

nachtmaar committed:
Use getEnvDefault function to reduce variables and code

santiago-ventura committed:
Add Build validation for getting function status

nachtmaar committed:
Enable loggger in tests and change StartTestManager according to upstream status

nachtmaar committed:
Add missing CRDs

nachtmaar committed:
Add todo

nachtmaar committed:
First working version of TestReconcile test

nachtmaar committed:
Add forgotten return statement in serveFunction

nachtmaar committed:
Remoev unused config map for Dockerfile

nachtmaar committed:
Add testing of build and buildTemplate

nachtmaar committed:
Make build template generic

nachtmaar committed:
Add test for first function config map

nachtmaar committed:
Fix race condition

nachtmaar committed:
Do not assume any order in parameters of build template

santiago-ventura committed:
Adding function name to logs in function_controller.go

santiago-ventura committed:
Rewrite getFunctionCondition. Update CRDs, Update to knative release 0.6.1

santiago-ventura committed:
Change test to use Service instead of Build

nachtmaar committed:
Add serving and build apis to test suite

santiago-ventura committed:
change test for getFunctionCondition

santiago-ventura committed:
Fix status template nil in function controller

nachtmaar committed:
Fix bug where empty resource was deleted

santiago-ventura committed:
Include new TestFunctionConditionServiceError

santiago-ventura committed:
getFunctionCondition returns if service is nil

nachtmaar committed:
Fix test TestFunctionConditionServiceError

santiago-ventura committed:
remove test TestGetFunctionControllerConfiguration and TestReconcileErrors

santiago-ventura committed:
Fix TestGetServiceSpec. Clean some code.

k15r commented on config/crds/kr6.0_build_v1alpha1_300-build.yaml:
```diff
@@ -0,0 +1,44 @@
+# Copyright 2018 The Knative Authors
```
rly? The header should be changed

venturasr commented:
Related issue: https://github.com/kyma-project/console/issues/836

nachtmaar commented on README.md:
```diff
@@ -98,29 +98,38 @@ curl -L https://raw.githubusercontent.com/knative/serving/v0.5.2/third_party/ist
-  --filename https://github.com/knative/build/releases/download/v0.5.0/build.yaml \
-  --filename https://github.com/knative/serving/releases/download/v0.5.2/monitoring.yaml \
-  --filename https://raw.githubusercontent.com/knative/serving/v0.5.2/third_party/config/build/clusterrole.yaml
+kubectl apply --selector knative.dev/crd-install=true \
```
We have version 0.6.1 in `Gopkg.toml`. Should we use CRDs with the same version?

venturasr replied:
I will replaced by  0.6.1

nachtmaar commented on README.md:
````diff
@@ -151,19 +160,34 @@ Then run the following commands:
+
 ```bash
-kubectl apply -f config/samples/runtime_v1alpha1_function.yaml
+kubectl apply -f config/samples/runtime_v1alpha1_function.yaml -n REPLACE_BY_NAMESPACE
````
Can we use `{}` or `<>` to indicate a variable? e.g. `{NAMESPACE}`

nachtmaar replied:
Same for all `REPLACE_BY_NAMESPACE` occurrences

venturasr replied:
I will use {NAMESPACE}

santiago-ventura committed:
removing useless comments in code and Removing references to images

nachtmaar commented on README.md:
```diff
@@ -98,29 +98,38 @@ curl -L https://raw.githubusercontent.com/knative/serving/v0.5.2/third_party/ist
-  --filename https://github.com/knative/build/releases/download/v0.5.0/build.yaml \
-  --filename https://github.com/knative/serving/releases/download/v0.5.2/monitoring.yaml \
-  --filename https://raw.githubusercontent.com/knative/serving/v0.5.2/third_party/config/build/clusterrole.yaml
+kubectl apply --selector knative.dev/crd-install=true \
```
I did not try out with `Minikube`, but for `GKE` this works with version `0.6.1`

```bash
kubectl apply --selector knative.dev/crd-install=true \
--filename https://github.com/knative/serving/releases/download/v0.6.1/serving.yaml \
--filename https://github.com/knative/build/releases/download/v0.6.0/build.yaml \
--filename https://github.com/knative/eventing/releases/download/v0.6.1/release.yaml \
--filename https://github.com/knative/eventing-sources/releases/download/v0.6.0/eventing-sources.yaml \
--filename https://github.com/knative/serving/releases/download/v0.6.1/monitoring.yaml \
--filename https://raw.githubusercontent.com/knative/serving/v0.6.1/third_party/config/build/clusterrole.yaml
```

```bash
kubectl apply --filename https://github.com/knative/serving/releases/download/v0.6.1/serving.yaml --selector networking.knative.dev/certificate-provider!=cert-manager \
--filename https://github.com/knative/build/releases/download/v0.6.0/build.yaml \
--filename https://github.com/knative/eventing/releases/download/v0.6.1/release.yaml \
--filename https://github.com/knative/eventing-sources/releases/download/v0.6.0/eventing-sources.yaml \
--filename https://github.com/knative/serving/releases/download/v0.6.1/monitoring.yaml \
--filename https://raw.githubusercontent.com/knative/serving/v0.6.1/third_party/config/build/clusterrole.yaml
```

Since we use the version `0.6.1` in `Gopkg.toml`, we should also use the CRDS with the same version

venturasr replied:
right!

venturasr replied:
README have been modified.

nachtmaar commented on config/crds/kr6.0_serving_v1alpha1_300-clusteringress.yaml:
```diff
@@ -0,0 +1,42 @@
+# Copyright 2018 The Knative Authors
```
This file is not required for the tests.

venturasr replied:
removed

nachtmaar commented on config/crds/kr6.0_build_v1alpha1_300-build.yaml:
```diff
@@ -0,0 +1,44 @@
+# Copyright 2018 The Knative Authors
```
I found the same CRD in `./vendor/github.com/knative/build/config/300-build.yaml`. Maybe we can add a symlink in `config/crd` to it ?

Probably we also have the other CRDs in the `vendor` directory

nachtmaar replied:
@k15r found out we can also directly reference it from the test suite like [this](https://github.com/ryanbrainard/kubebuilder-p2/blob/4035be7426682c510990779cfb4f4d492921b201/pkg/controller/artist/artist_controller_suite_test.go#L40). Probably better that the symlink

venturasr replied:
Following @k15r  suggestions combined with symlink

k15r commented on Makefile:
```diff
@@ -16,8 +16,8 @@ manager: generate fmt vet
-run: generate fmt vet
-	go run ./cmd/manager/main.go
+#run: generate fmt vet
+#	go run ./cmd/manager/main.go
```
why?
either remove or activate please

venturasr replied:
It is activated

nachtmaar committed:
Polishing

nachtmaar committed:
Merge branch 'refactored-runtime' of github.com:kyma-incubator/runtime into refactored-runtime

nachtmaar committed:
Fix manager image patches

nachtmaar committed:
Use "Hello World" instead of "Hello World2"

nachtmaar committed:
Remove sed backup file

nachtmaar committed:
Revert config.yaml

nachtmaar commented:
When having a build with error status, the function status is still `Deploying` but should be in error state. I will add a test for it

k15r commented on pkg/controller/function/function_controller.go:
```diff
@@ -70,6 +72,28 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	}
+
+	// Watch for changes to Service
+	err = c.Watch(&source.Kind{Type: &servingv1alpha1.Service{}}, &handler.EnqueueRequestForObject{})
```
Are we sure that we enqueue the reconcilerequest for the correct object? imho we need to enqueue for the owner of the source as the created reconcile request contains the name of the object. so if build has a different name than the function, we will not successfully update the status of the function as the function cannot be found

venturasr replied:
Actually, I think there is no need for watching the Build. When I started to have problem with the update of serving status I started to watch the Build and the Route and Configuration to try to figure out the status. But once I updated the serving crds the problem with the Service status got solved.

venturasr replied:
I think the controller should watch the function and the service.

nachtmaar replied:
What happens if the build encounters an error, e.g. due to wrong docker registry credentials?
Does the service also provide the error or do we need to watch the build?

venturasr replied:
Currently it watching Function, Build and Serving.

k15r replied:
imho we need to NOT delete the build object before queueing a new build. storing the old build will enable us to provide some sort of audit data on the build process.
This means, that we cannot have the same name on the build object and the function object, which in return means we need to enqueue the request for the owner.

e.g. 
```
err = c.Watch(&source.Kind{Type: &buildv1alpha1.Build{}}, &handler.EnqueueRequestForOwner{
		OwnerType: &runtimev1alpha1.Function{},
		IsController: true,
	})
```

k15r commented on pkg/controller/function/function_controller.go:
```diff
@@ -83,62 +107,178 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-	fnConfigNamespaceEnv := os.Getenv("CONTROLLER_CONFIGMAP_NS")
+	// Get Function Controller Configuration
+	fnConfig := &corev1.ConfigMap{}
+	if err := r.getFunctionControllerConfiguration(fnConfig); err != nil {
```
it should be possible that the reconcile function does not have to read the controller configuration. In a perfect world all user input has already been validated by the webhook, so whatever reaches here is already enriched to contain all necessary information to create a running function.
the reconcile function could the focus on creating objects and updating status

venturasr replied:
I think the only reason why is being read is because the RuntimeInfo. We need to know the ServiceAccountName in order to create the Build.

https://github.com/kyma-incubator/runtime/blob/695c0b65d2aa8121404ebbdd89883bcfce06e410/pkg/utils/controller_utils.go#L46

nachtmaar replied:
Created a follow-up issue for that: https://github.com/kyma-project/kyma/issues/4826

k15r commented on pkg/apis/runtime/v1alpha1/function_types.go:
```diff
@@ -45,15 +45,35 @@ type FunctionSpec struct {
+	// Indicates that function has an error condition.
+	FunctionConditionError FunctionCondition = "Error"
+	// Indicates that function has a deploying condition.
+	FunctionConditionDeploying FunctionCondition = "Deploying"
```
Should we add a condition for Building?

venturasr replied:
We were thinking more in "Unkown" condition. I am not sure bout Build condition for the function. A build is happening underneath but I think is more a knative stuff. However we can include it `Unknown, Build, Deploying, Running, Updating and Error`

nachtmaar commented on pkg/apis/runtime/v1alpha1/function_types.go:
```diff
@@ -45,15 +45,35 @@ type FunctionSpec struct {
+type FunctionCondition string
+
+const (
+	// Indicates that function has a running condition.
```
What about having something like "unknown" or "N/A" status by default? Currently, then you create a function it is "" which feels like the status is broken. So use "unknown" or "N/A" as default status

venturasr replied:
"unknown" it is a good one

santiago-ventura committed:
Replace real crds for symlinks to be used by the test. Organize imports

nachtmaar commented on pkg/controller/function/function_controller_test.go:
```diff
@@ -218,106 +228,135 @@ func TestReconcile(t *testing.T) {
+	g.Expect(ksvcUpdated.Spec.ConfigurationSpec.Template.Spec.RevisionSpec.PodSpec.Containers[0].Image).
 		To(gomega.Equal(fmt.Sprintf("test/%s-%s:%s", "default", "foo", functionSha)))
+
+	//g.Expect(fnUpdatedFetched.Status.Condition).To(gomega.Equal(runtimev1alpha1.FunctionConditionRunning))
```
Missing check of function condition. Should be uncommented

k15r commented on pkg/apis/runtime/v1alpha1/function_types.go:
```diff
@@ -45,15 +45,35 @@ type FunctionSpec struct {
+
 // FunctionStatus defines the observed state of Function
 type FunctionStatus struct {
+	Condition FunctionCondition `json:"condition,omitempty"`
```
imho a list of Conditions would be nice here, so you could tell the user on a finer level whats currently happening to the function

see basically every knative CRD e.g. https://knative.dev/docs/reference/serving-api/#github.com/knative/serving/pkg/apis/serving/v1alpha1.ServiceStatus

venturasr replied:
It will be manage in another issue.

k15r commented on pkg/controller/function/function_controller.go:
```diff
@@ -83,62 +107,178 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+	// Create function's image name
+	hash := sha256.New()
+	print("foundCM: %v", foundCm)
```
debug message ... please log as debug

venturasr replied:
I will removed. That was a flag for debugging.

santiago-ventura committed:
Update README base on comments in PR

k15r commented on pkg/controller/function/function_controller.go:
```diff
@@ -148,49 +288,251 @@ func (r *ReconcileFunction) Reconcile(request reconcile.Request) (reconcile.Resu
 
-	dockerRegistry := rnInfo.RegistryInfo
-	imageName := fmt.Sprintf("%s/%s-%s:%s", dockerRegistry, fn.Namespace, fn.Name, functionSha)
+	// Check if the Build instance can be get it.
```
? i'm sorry, but what should this mean?

venturasr replied:
It is just a check after create the Build, to see if can be retrieve it without error. Probably is not needed. Should be removed?

k15r reviewed (commented):
The comments should be implemented in later PRs... so this will just be a comment without requesting changes

nachtmaar commented on pkg/controller/function/function_controller.go:
```diff
@@ -83,62 +107,178 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+		return reconcile.Result{}, err
+	}
+	log.Info("build Function image [done]", "function_name:", fn.Name)
```
We should remove all logs with the pattern `".* [done]"`. It was meant for debugging only and bloates the log output probably to much.

venturasr replied:
Yes, it is not needed. If there is an error this will happens when the Build is being created.

santiago-ventura committed:
Add Building and Serving status

santiago-ventura committed:
Handle errors from updateFunctionStatus

santiago-ventura committed:
removing no needed logs

santiago-ventura committed:
Handle cluster nil in Build Status

santiago-ventura committed:
Handle cluster nil in Build Status

santiago-ventura committed:
Remove logs used for debugging

santiago-ventura committed:
Remove deprecated domain and revert to FunctionConditionDeploying in getFunctionCondition

nachtmaar committed:
Handle build condition error

nachtmaar committed:
Add and fix status tests

nachtmaar committed:
Merge branch 'refactored-runtime' of github.com:kyma-incubator/runtime into refactored-runtime

nachtmaar committed:
Fix tests after merge

santiago-ventura committed:
Handle error when function is deleted

nachtmaar commented on pkg/controller/function/function_controller.go:
```diff
@@ -83,62 +99,173 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+		return err
+	}
+
+	// TODO REMOVE LOG
```
TODO

nachtmaar commented on pkg/controller/function/function_controller.go:
```diff
@@ -148,49 +275,244 @@ func (r *ReconcileFunction) Reconcile(request reconcile.Request) (reconcile.Resu
 		foundCm.Data = deployCm.Data
-		log.Info("Updating ConfigMap", "namespace", deployCm.Namespace, "name", deployCm.Name)
-		err = r.Update(context.TODO(), foundCm)
+		// TODO: why is this required ??
```
We should remove all added todos unless there is a follow up story for that with a issue number referenced

santiago-ventura committed:
Add test TestCreateFunctionCreatedAndDeleted

nachtmaar commented on Dockerfile:
```diff
@@ -14,4 +14,5 @@ RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -a -o manager github.com/kyma
 FROM ubuntu:latest
 WORKDIR /
 COPY --from=builder /go/src/github.com/kyma-incubator/runtime/manager .
+# COPY --from=builder /go/ /go/
```
Debug leftover. Should be removed

nachtmaar commented on README.md:
```diff
@@ -98,29 +98,38 @@ curl -L https://raw.githubusercontent.com/knative/serving/v0.5.2/third_party/ist
 
 #### Manager running locally
 
+modify config/config.yaml to include your docker.io credentials (base64 encoded) and update the dockerregistry value to your docker.io username
```
"dockerregistry" => "docker registry"

santiago-ventura committed:
Remove no needed tests

nachtmaar committed:
Add TestReconcileDeleteFunction test
- cleanup created resources in each test
- ensure no reconcile errors occur

santiago-ventura committed:
Removed no needed comments

santiago-ventura committed:
Remove TODOs and no needed comments.

k15r commented on pkg/controller/function/function_controller.go:
```diff
@@ -83,62 +99,171 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+			return reconcile.Result{}, nil
+		}
+		// status of the functon must change to error.
+		r.updateFunctionStatus(fn, runtimev1alpha1.FunctionConditionError)
```
for something like this it would make sense to implement a reason field in the status subressource ...

venturasr replied:
True

k15r commented on pkg/controller/function/function_controller.go:
```diff
@@ -83,62 +99,171 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create function's image name
+	hash := sha256.New()
+	hash.Write([]byte(foundCm.Data["handler.js"] + foundCm.Data["package.json"]))
+	functionSha := fmt.Sprintf("%x", hash.Sum(nil))
```
we could use this functionsha in the build-objects name as well...

venturasr replied:
Now it is taking the functionSha and generating an short sha from the slice.

nachtmaar committed:
Remove commented code

nachtmaar committed:
Merge branch 'refactored-runtime' of github.com:kyma-incubator/runtime into refactored-runtime

santiago-ventura committed:
Generate unique build name. The Build is watched is owned by a Function

santiago-ventura committed:
Make short sha shorter

santiago-ventura committed:
Make short sha shorter

santiago-ventura committed:
Seeting test timeout to 20 seconds. Making imagePullPolicy as Never in manager_image_patch.yaml

k15r commented on pkg/utils/build_utils.go:
```diff
@@ -0,0 +1,179 @@
+	Timeout            time.Duration
+}
+
+var dockerFileName = map[string]string{
```
why is this hardcoded

nachtmaar committed:
Remove todo comment

nachtmaar committed:
Merge branch 'refactored-runtime' of github.com:kyma-incubator/runtime into refactored-runtime

nachtmaar reviewed (approved):
LGTM

nachtmaar commented:
@venturasr please use `squash and merge` if possible :)

santiago-ventura committed:
Remove BuildUtil struct. change manager image patch yamls

santiago-ventura committed:
Remover comments with BuildUtils

santiago-ventura committed:
Solve conflicts

santiago-ventura committed:
Fix wrong buildName issue in function_controller_test.go
